collections/src/main/scala/strawman/collection/ArrayOps.scala

```diff
@@ -1,34 +1,47 @@
 package strawman
 package collection
 
-import scala.{AnyVal, Array, ArrayIndexOutOfBoundsException, Char, Int, throws}
+import scala.{AnyVal, Array, ArrayIndexOutOfBoundsException, Int, throws, Boolean, Serializable, Unit, `inline`}
 import scala.Predef.???
 import mutable.{ArrayBuffer, GrowableBuilder}
-
+import immutable.ImmutableArray
 import scala.reflect.ClassTag
 
-class ArrayOps[A](val xs: Array[A])
-  extends AnyVal
-    with IterableOnce[A]
-    with IndexedSeqOps[A, immutable.IndexedSeq, Array[A]]
-    with StrictOptimizedIterableOps[A, Seq, Array[A]]
-    with ArrayLike[A] {
+object ArrayOps {
+  class WithFilter[A](p: A => Boolean, ao: ArrayOps[A]) extends collection.WithFilter[A, immutable.IndexedSeq] {
+    protected[this] def filtered = View.Filter(ao.toIterable, p, isFlipped = false)
+    def map[B](f: A => B): immutable.IndexedSeq[B] = ao.iterableFactory.from(View.Map(filtered, f))
+    def flatMap[B](f: A => IterableOnce[B]): immutable.IndexedSeq[B] = ao.iterableFactory.from(View.FlatMap(filtered, f))
+    def foreach[U](f: A => U): Unit = filtered.foreach(f)
+    def map[B: ClassTag](f: A => B): Array[B] = ao.fromTaggedIterable(View.Map(filtered, f))
+    def flatMap[B: ClassTag](f: A => IterableOnce[B]): Array[B] = ao.fromTaggedIterable(View.FlatMap(filtered, f))
+    def withFilter(q: A => Boolean): WithFilter[A] = new WithFilter[A](a => p(a) && q(a), ao)
+  }
+}
+
+class ArrayOps[A](val xs: Array[A]) extends AnyVal
+  with IterableOnce[A]
+  with IndexedSeqOps[A, immutable.IndexedSeq, Array[A]]
+  with StrictOptimizedIterableOps[A, Seq, Array[A]]
+  with ArrayLike[A] {
+
+  protected def fromTaggedIterable[B: ClassTag](coll: Iterable[B]): Array[B] = coll.toArray[B]
+
+  // Note: Declaring the return type as ArrayOps.WithFilter[A] breaks Dotty 0.4.0-RC1
+  override def withFilter(p: A => Boolean): collection.WithFilter[A, immutable.IndexedSeq] = new ArrayOps.WithFilter[A](p, this)
 
-  def toIterable = ArrayView(xs)
+  def toIterable: ImmutableArray[A] = ImmutableArray.wrapArray(xs)
```

Comment:
Actually I think it is not safe to expose it as an `immutable` one because the underlying array `xs` is still available and can be mutated.

Reply:
That's a general risk with `ImmutableArray` but it is unavoidable if we want to make the default `Seq` a `collection.immutable.Seq` and support varargs as `Seq`.

Or should we limit the implicit conversion from `Array` to produce a generic `Seq`? Varargs could still expose it as an immutable one. It may still be somewhat confusing once we make the immutable `Seq` the default one.

Reply:
Yes, I would expose `collection.IndexedSeq` as the public result type of `toIterable`. We can still safely use `ImmutableArray` to support varargs since the reference to the underlying mutable `Array` will not leak (or will it?).

Reply:
It will but I'd classify that as an abuse of varargs. If you mutate an array after passing it to a varags method all bets are off.